python/ray/actor.py

```diff
@@ -93,73 +100,123 @@ def get_actor_checkpoint(worker, actor_id):
-            # Execute the checkpoint task.
-            actor_checkpoint_failed, error = method(actor, *args)
-            # If the checkpoint was successfully loaded, update the actor's
-            # task counter and set a flag to notify the local scheduler, so
-            # that the task following the checkpoint can run.
-            if not actor_checkpoint_failed:
-                worker.actor_task_counter = task_counter + 1
-                # Once the actor has resumed from a checkpoint, it counts as
-                # loaded.
-                worker.actor_loaded = True
-            # Report to the local scheduler whether this task succeeded in
-            # loading the checkpoint.
-            worker.actor_checkpoint_failed = actor_checkpoint_failed
-            # If there was an exception during the checkpoint method, re-raise
-            # it after updating the actor's internal state.
-            if error is not None:
-                raise error
-            return None
-        else:
-            # Update the worker's internal state before executing the method in
-            # case the method throws an exception.
-            worker.actor_task_counter = task_counter + 1
-            # Once the actor executes a task, it counts as loaded.
-            worker.actor_loaded = True
-            # Execute the actor method.
-            return method(actor, *args)
+    def actor_method_executor(dummy_return_id, actor, *args):
+        # Update the actor's task counter to reflect the task we're about to
+        # execute.
+        worker.actor_task_counter += 1
+
+        # If this is the first task to execute on the actor, try to resume from
+        # a checkpoint.
+        if actor_imported and worker.actor_task_counter == 1:
+            checkpoint_resumed = restore_and_log_checkpoint(worker, actor)
+            if checkpoint_resumed:
+                # NOTE(swang): Since we did not actually execute the __init__
+                # method, this will put None as the return value. If the
+                # __init__ method is supposed to return multiple values, an
+                # exception will be logged.
+                return
+
+        # Execute the assigned method.
+        method_error = None
+        try:
+            method_returns = method(actor, *args)
+        except Exception as e:
+            method_error = e
+
+        # Check whether we should checkpoint the actor.
+        if (actor_imported and worker.actor_checkpoint_interval > 0):
+            # The user has set a checkpointing interval.
+            if (worker.actor_task_counter % worker.actor_checkpoint_interval ==
+                    0 and method_name != "__ray_checkpoint__"):
+                # We've executed checkpoint_interval tasks since the last
+                # checkpoint and the last method executed was not a checkpoint.
+                save_and_log_checkpoint(worker, actor)
+
+        if method_error is not None:
+            raise method_error
```

Comment:
does this preserve the exception backtrace when we reraise the exception here? I can give this a try to see

Reply:
Nice catch! It doesn't; we have to re-raise within the `except` block to preserve the backtrace.